File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

void dfs(struct TreeNode* root, int *result, int max_val) {
    int current_max;
    
    if (!root) {
        return;
    }
    
    current_max = max_val;
    if (root->val >= max_val) {
        (*result)++;
        current_max = root->val;
    }
    
    dfs(root->left, result, current_max);
    dfs(root->right, result, current_max);
}

int goodNodes(struct TreeNode* root){
    int result = 0;
    
    dfs(root, &result, INT_MIN);
    
    return result;
}

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        
        result = 0
        
        def dfs(root: TreeNode, maxVal: int):
            nonlocal result
            
            if not root:
                return
            
            curMax = maxVal
            if root.val >= maxVal:
                result += 1
                curMax = root.val
                
            dfs(root.left, curMax)
            dfs(root.right, curMax)
            
        
        dfs(root, float('-inf'))
        
        return result
            
            
                